examples/runners/ash/src/main.rs

```diff
@@ -1,259 +1,229 @@
-// This file is copied pretty directly from
-// https://github.com/MaikKlein/ash/blob/master/examples/src/bin/triangle.rs
-// with the only major differences being
-// 1) the minimum vulkan version bumped from v1.0 to v1.1
-// 2) the path to the vertex shader is env!("example_shader.spv"), i.e.
-// let vertex_shader_file = include_bytes!(env!("example_shader.spv"));
-
-use ash::extensions::ext::DebugUtils;
-use ash::extensions::khr::{Surface, Swapchain};
-use ash::util::*;
-use ash::version::{DeviceV1_0, EntryV1_0, InstanceV1_0};
-use ash::{vk, Device, Instance};
-use shared::ShaderConstants;
-use std::borrow::Cow;
-use std::default::Default;
-use std::ffi::{CStr, CString};
-use std::io::Cursor;
-use std::ops::Drop;
+use ash::{
+    extensions::{ext, khr},
+    util::read_spv,
+    version::{DeviceV1_0, EntryV1_0, InstanceV1_0},
+    vk,
+};
+
+use winit::{
+    event::{Event, VirtualKeyCode, WindowEvent},
+    event_loop::{ControlFlow, EventLoop},
+};
+
+use serde::Deserialize;
+
+use std::{
+    borrow::Cow,
+    collections::HashMap,
+    default::Default,
+    ffi::{CStr, CString},
+    fs::File,
+    ops::Drop,
+    path::PathBuf,
+    process::{Command, Stdio},
+    sync::atomic::{AtomicBool, Ordering},
+};
+
 use structopt::StructOpt;
 
-fn shader_module() -> &'static [u8] {
-    &include_bytes!(env!("sky_shader.spv"))[..]
+#[derive(Copy, Clone)]
+pub struct ShaderConstants {
+    pub width: u32,
+    pub height: u32,
 }
 
-// Simple offset_of macro akin to C++ offsetof
-#[macro_export]
-macro_rules! offset_of {
-    ($base:path, $field:ident) => {{
-        #[allow(unused_unsafe)]
-        unsafe {
-            let b: $base = mem::zeroed();
-            (&b.$field as *const _ as isize) - (&b as *const _ as isize)
-        }
-    }};
+#[derive(Debug, StructOpt)]
+#[structopt()]
+pub struct Options {
+    /// Use Vulkan debug layer (requires Vulkan SDK installed)
+    #[structopt(short, long)]
+    debug_layer: bool,
 }
 
-unsafe fn any_as_u8_slice<T: Sized>(p: &T) -> &[u8] {
-    ::std::slice::from_raw_parts((p as *const T) as *const u8, ::std::mem::size_of::<T>())
-}
+// This is not an ideal solution, but it's simple and doesn't require an async runtime.
+static NEEDS_REBUILD: AtomicBool = AtomicBool::new(false);
+static IS_COMPILING: AtomicBool = AtomicBool::new(false);
+static mut NEW_SHADERS: Vec<SpirvShader> = Vec::<SpirvShader>::new();
```

Comment:
While the logic you wrote below seems like it could probably work correctly in this case; I think this kind of state management through atomic booleans is ripe for potential race conditions down the line. Would it be an option to re-do this part of the logic to use an mpsc channel instead to keep it a bit more tidy and less likely to have threading bugs?